### PR TITLE
Add ReorderedPartitionsRDD

### PR DESCRIPTION
**Pull request by tpoterba**
This allows us to coerce as sorted 'files.chr*.vcf',
no matter the order of the files in globbing.

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 1]
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
...
+      r
+    }
 
     if (partitionsSorted) {
-      val keyInfo = keyInfoOption.get
-      val partitioner = OrderedPartitioner(keyInfo.init.map(pi => pi.max))
-      val sortedness = keyInfo.map(_.sortedness).min
+      val partitioner = OrderedPartitioner[PK, K](sortedKeyInfo.init.map(_.max))
+      val sortedness = sortedKeyInfo.map(_.sortedness).min
+      val scrambledRDD = new ReorderedPartitionsRDD[(K, V)](rdd, sortedKeyInfo.map(_.partIndex))
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: I'd changed `scrambled` to `reordered` or something.  Scrambled sounds like you're creating chaos, but you're creating order.

File: src/main/scala/org/broadinstitute/hail/sparkextras/ReorderedPartitionsRDD.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package org.broadinstitute.hail.sparkextras
+
+import org.apache.spark.{Dependency, NarrowDependency, Partition, TaskContext}
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+case class ReorderedPartition(index: Int, oldPartition: Partition) extends Partition
+
+class ReorderedPartitionsRDD[T](prev: RDD[T], oldIndices: Array[Int])(implicit tct: ClassTag[T])
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/ReorderedPartitionsRDD.scala]: `prev` should be `@transient` and you should implement clearDependencies.

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 3]
```diff
@@ -75,15 +75,15 @@ class OrderedRDDSuite extends SparkSuite {
...
     }
 
     property("locusSorted") = Prop.forAll(locusSorted) { case (nPar, s) =>
-      check(sc.parallelize(s, nPar).toOrderedRDD[Locus])
+      check(sc.parallelize(s, nPar).toOrderedRDD)
     }
 
     property("fullySorted") = Prop.forAll(sorted) { case (nPar, s) =>
-      check(sc.parallelize(s, nPar).toOrderedRDD[Locus])
+      check(sc.parallelize(s, nPar).toOrderedRDD)
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: Should you call coerce and verify the status for `sorted` and `locusSorted`?  Might also have an empty multi-partition one.

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 4]
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
         .collect()
         .forall { case (v1, v2) => v1 == v2 }
     }
+
+    property("partitionReordering") = Prop.forAll(sorted) { case (nPar, is) =>
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: Should you run this against `sorted`, `locusSorted` or `random`?

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 5]
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
...
+
+      val rdd = sc.parallelize(is, nPar)
+      val shuffledIndices = Array.range(0, nPar).sortBy(_ => scala.util.Random.nextDouble())
+      val scrambled = new ReorderedPartitionsRDD(rdd, shuffledIndices)
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: Should we add `RichRDD.reorderPartitions(oldIndices: Array[Int])`?

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 6]
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
...
+
+      val (status, result) = OrderedRDD.coerce(rdd)
+      val (status2, result2) = OrderedRDD.coerce(scrambled)
+
+
+      status == status2 && result.fullOuterJoin(result2)
+        .collect()
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: You can do `forall` directly against the RDDs.

File: src/main/scala/org/broadinstitute/hail/sparkextras/ReorderedPartitionsRDD.scala [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package org.broadinstitute.hail.sparkextras
+
+import org.apache.spark.{Dependency, NarrowDependency, Partition, TaskContext}
+import org.apache.spark.rdd.RDD
+
+import scala.reflect.ClassTag
+
+case class ReorderedPartition(index: Int, oldPartition: Partition) extends Partition
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/ReorderedPartitionsRDD.scala]: `ReorderedPartitionsRDDPartition` is traditional.

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 8]
```diff
@@ -151,6 +151,22 @@ class OrderedRDDSuite extends SparkSuite {
         .collect()
         .forall { case (v1, v2) => v1 == v2 }
     }
+
+    property("partitionReordering") = Prop.forAll(sorted) { case (nPar, is) =>
+
+      val rdd = sc.parallelize(is, nPar)
+      val shuffledIndices = Array.range(0, nPar).sortBy(_ => scala.util.Random.nextDouble())
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: This is a annoying change since the Prop/Gen syntax is so bad, but it would be natural to have `Gen.shuffle(Gen.const(0 until nPar))` as another generator item.

**Comment by cseed** [comment]
Back to you.

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala (hunk 4 above)
Comment (tpoterba) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: `random` will end up doing a shuffle.  I'll choose from sorted or locusSorted.

File: src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    val locationAvail = range.flatMap(i =>
+      prev.preferredLocations(prevPartitions(i)))
+      .groupBy(identity)
+      .mapValues(_.length)
+
+    if (locationAvail.isEmpty)
+      return Seq.empty
+
+    val m = locationAvail.values.max
+    locationAvail.filter(_._2 == m)
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala]: I feel like the standard library should (but seemingly doesn't) have a

```scala
trait Seq[+A] {
  def maxesBy[C[_], B](f: (A) => B)(implicit cbf: CanBuildFrom[Nothing, B, C[B]]): C[B] = {
    var c = cbf()
    if (this.isEmpty)
      c.result
    else {
      var biggest = this.head
      c += biggest
      for (a <- this.tail) {
        val b = f(a)
        if (b > biggest) {
          c = cbf()
          biggest = b
          c += biggest
        } else if (b == biggest) {
          c += b
        }
      }
      c.result
    }
  }
}
```

File: src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  override def getPreferredLocations(partition: Partition): Seq[String] = {
+    val prevPartitions = prev.partitions
+    val range = partition.asInstanceOf[BlockedRDDPartition].range
+
+    val locationAvail = range.flatMap(i =>
+      prev.preferredLocations(prevPartitions(i)))
+      .groupBy(identity)
+      .mapValues(_.length)
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala]: would
```scala
range.flatMap(i => prev.preferredLocations(prevPartitions(i)).map((_, _.length)))
  .toMap
```
be more explicit about what's going on? Not sure if the performance is any different.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 11]
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
...
-      keyInfoOption.exists(keyInfo =>
-        keyInfo.zip(keyInfo.tail).forall { case (p, pnext) =>
-          val r = p.max < pnext.min
-          if (!r)
-            log.info(s"not sorted: p = $p, pnext = $pnext")
-          r
-        })
+    val keyInfo = keys.mapPartitionsWithIndex { case (i, it) =>
+      if (it.hasNext)
+        Iterator(PartitionKeyInfo.apply(i, it))
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: unnecessary `.apply`, I think?

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 12]
```diff
@@ -2,20 +2,26 @@ package org.broadinstitute.hail.sparkextras
...
 import scala.collection.mutable
 import scala.collection.mutable.ArrayBuffer
+import scala.language.existentials
 import scala.reflect.ClassTag
 import scala.util.hashing._
-import scala.language.existentials
 
 object OrderedRDD {
 
+  final val ORDERED_PARTITIONER = 0
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: might be worth a

```scala
type COERCION_METHOD = Int
```
too

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 13]
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
...
+        log.info(s"not sorted: p = $p, pnext = $pnext")
+      r
+    }
 
     if (partitionsSorted) {
-      val keyInfo = keyInfoOption.get
-      val partitioner = OrderedPartitioner(keyInfo.init.map(pi => pi.max))
-      val sortedness = keyInfo.map(_.sortedness).min
+      val partitioner = OrderedPartitioner[PK, K](sortedKeyInfo.init.map(_.max))
+      val sortedness = sortedKeyInfo.map(_.sortedness).min
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Is it usually the case that every partition has the same sortedness? In particular I'm wondering if you ever have one unsorted partition and the rest sorted.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 14]
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
...
+    (implicit kOk: OrderedKey[PK, K]): (Int, OrderedRDD[PK, K, V]) = {
     import kOk.kct
     import kOk.pkct
     import kOk.kOrd
     import kOk.pkOrd
     import Ordering.Implicits._
 
     if (rdd.partitions.isEmpty)
-      return empty(rdd.sparkContext)
+      return (ORDERED_PARTITIONER, empty(rdd.sparkContext))
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Is this `AS_IS`?

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala [hunk 15]
```diff
@@ -151,10 +158,29 @@ class OrderedRDDSuite extends SparkSuite {
...
+      newPartitions <- Gen.shuffle(rdd.partitions.indices)
+    ) yield (rdd, rdd.reorderPartitions(newPartitions.toArray))
+
+    property("partitionReordering") = Prop.forAll(scrambled) { case (original, scrambled) =>
+
+      val (status, result) = OrderedRDD.coerce(original)
+      val (status2, result2) = OrderedRDD.coerce(scrambled)
+
+      status == status2 && result.fullOuterJoin(result2)
+        .forall { case (v, (s1, s2)) => s1 == s2 }
```
Comment (danking) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: This makes me wish the quick check library had a way to point out what lead to the falsification of the property. Here we really have a huge number of conjunctions, and it might not be obvious which one is causing the problem.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 16]
```diff
@@ -38,72 +44,80 @@ object OrderedRDD {
   }
 
   def apply[PK, K, V](rdd: RDD[(K, V)], fastKeys: Option[RDD[K]] = None)(implicit kOk: OrderedKey[PK, K]): OrderedRDD[PK, K, V] = {
+    val (_, orderedRDD) = coerce(rdd, fastKeys)
+    orderedRDD
+  }
+
+  def coerce[PK, K, V](rdd: RDD[(K, V)], fastKeys: Option[RDD[K]] = None)
+    (implicit kOk: OrderedKey[PK, K]): (Int, OrderedRDD[PK, K, V]) = {
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: It seems a lot of places toss out the first element of the pair, I wonder if there should be a separate method for those places that does the throwing away itself? Keeps the API slimmer.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala (hunk 14 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: I'm not sure.  It doesn't really matter, I guess.  I think I used AS_IS to indicate that you spent an action to check, but then you're done  The ORDERED_PARTITIONER really means that no work was done, which happens when the partitioner is ordered or the RDD is empty

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala (hunk 16 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: That's what the `apply` above is for.  This lets us test freely (and maybe even helps when the query optimizer comes to town) but the previous API doesn't change.

Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Ah yep, totally missed this.

**Comment by tpoterba** [comment]
Addressed comments, ready for another look.  Change to `Gen.partition` revealed a few other bugs in master, which don't have elegant solutions:

1.  orderedPartitioner either needs an `empty` boolean flag, or a special `EmptyOrderedPartitioner` where `partitioner.nPartitions` is 0.  Otherwise, rdd.fullOuterJoin(other) crashes because it looks at partitioner nPartitions instead of rdd nPartitions.  This fix involves more logic in VSM.read as well.

2.  OrderedLeftJoinRDD needs to cache partitions and pass them to OrderedRDDIterator.  There is no guarantee that `getPartitions` can be called on an executor, and this was crashing in particular for ParallelCollectionRDDs.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala (hunk 13 above)
Comment (tpoterba) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: This isn't something we can assume.  In particular, if the dataset is locus-sorted but not variant-sorted, we would expect some partitions to be K-sorted and some to be T-sorted.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala [hunk 17]
```diff
@@ -13,22 +13,23 @@ import scala.reflect.{ClassTag, classTag}
...
+  (implicit val kOk: OrderedKey[PK, K])
   extends Partitioner {
 
   import kOk.pkct
   import kOk.pkOrd
   import Ordering.Implicits._
 
-  require(rangeBounds.isEmpty ||
-    rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => left < right })
+  require(!empty || rangeBounds.isEmpty)
```
Comment (danking) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala]: might be clearer with implication
```scala
import org.broadinstitute.hail.Utils.richBoolean
// ...
require(empty ==> rangeBounds.isEmpty)
```

**Comment by danking** [comment]
I suppose the more "object oriented" way would be to have two classes. I don't have strong feelings either way, I think it's fine as it is.

File: src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala (hunk 10 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala]: Two comments:

 - I don't think that's the same.  I want to count how many previous partitions prefer each location.  Hence the `groupBy`.

 - Syntactic note, each instance of `_` is a separate argument.  If you want to use an argument multiple times, you can't use the implicit lambda syntax.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala (hunk 17 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala]: Agreed, but I think by default that's only available in the test framework.  Or did we add it somewhere?

```
scala> true ==> false
<console>:8: error: value ==> is not a member of Boolean
              true ==> false
                   ^
```

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala [hunk 18]
```diff
@@ -13,22 +13,23 @@ import scala.reflect.{ClassTag, classTag}
 import scala.util.Random
 import scala.util.hashing.{MurmurHash3, byteswap32}
 
-case class OrderedPartitioner[PK, K](rangeBounds: Array[PK], ascending: Boolean = true)(implicit val kOk: OrderedKey[PK, K])
+case class OrderedPartitioner[PK, K](rangeBounds: Array[PK], ascending: Boolean = true, empty: Boolean = false)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala]: I'm worried about this `empty` business.  It seems error prone and I'm not convinced it is getting set correctly everywhere, e.g. `OrderedRDD.apply(rdd, ranges)`.  I think it should take a non-default `numPartitions: Int` argument.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedLeftJoinRDD.scala [hunk 19]
```diff
@@ -75,6 +76,8 @@ class OrderedLeftJoinRDD[PK, K, V1, V2](left: OrderedRDD[PK, K, V1], right: Orde
   extends RDD[(K, (V1, Option[V2]))](left.sparkContext, Seq(new OneToOneDependency(left),
     new OrderedDependency(left.orderedPartitioner, right.orderedPartitioner, right)): Seq[Dependency[_]]) {
 
+  val cachedRightPartitions = right.partitions
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedLeftJoinRDD.scala]: I'd just call this `rightPartitions` (and maybe make it private?)

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 20]
```diff
@@ -162,34 +163,30 @@ object VariantSampleMatrix {
...
-      } catch {
-        case _: InvalidClassException => None
-        case _: FileNotFoundException => None
       }
-    }
 
-    val orderedRDD = partitioner match {
-      case Some(p) =>
-        OrderedRDD[Locus, Variant, (Annotation, Iterable[Genotype])](readRDD(skipGenotypes), p)
+    val partitioner = try {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: The indentation is wrong here.

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 21]
```diff
@@ -706,17 +703,18 @@ class VariantSampleMatrix[T](val metadata: VariantMetadata,
...
-
-        (mapOp(va, gs.iterator
-          .zip(localSampleIdsBc.value.iterator
-            .zip(localSampleAnnotationsBc.value.iterator)).foldLeft(zeroValue) {
-          case (acc, (g, (s, sa))) =>
-            seqOp(acc, v, va, s, sa, g)
-        }), gs)
-      }.asOrderedRDD[Locus])
+      rdd = rdd.mapValuesWithKey {
+        case (v, (va, gs)) =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: Unnecessary change in indentation (I think?).  I've been seeing IntelliJ do this sporadically and I'm not sure what's up with that.

File: src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala [hunk 22]
```diff
@@ -1008,9 +1006,10 @@ class RichVDS(vds: VariantDataset) {
       val (newSignatures2, f2) = vds1.deleteVA("aIndex")
       vds1.copy(wasSplit = false,
         vaSignature = newSignatures2,
-        rdd = vds1.rdd.mapValuesWithKey { case (v, (va, gs)) =>
-          (f2(f1(va)), gs.lazyMap(g => g.copy(fakeRef = false)))
-        }.asOrderedRDD[Locus])
+        rdd = vds1.rdd.mapValuesWithKey {
+          case (v, (va, gs)) =>
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala]: Again.

File: src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala (hunk 15 above)
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/utils/OrderedRDDSuite.scala]: I agree, although I have no idea what to do about this.

File: src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala [hunk 23]
```diff
@@ -170,7 +170,7 @@ class VSMSuite extends SparkSuite {
   }
 
   @Test def testReadWrite() {
-    val p = forAll(VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random)) { (vsm: VariantSampleMatrix[Genotype]) =>
+    val p = forAll(VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random).filter(_.nVariants == 0)) { (vsm: VariantSampleMatrix[Genotype]) =>
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala]: `.filter(_.nVariants == 0)`?  Whoa, that can't be right.

File: src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala [hunk 24]
```diff
@@ -306,7 +306,7 @@ class VSMSuite extends SparkSuite {
 
   @Test def testCoalesce() {
     val g = for (
-      vsm <- VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random);
+      vsm <- VariantSampleMatrix.gen[Genotype](sc, VSMSubgen.random).filter(_.nPartitions > 0);
```
Comment (cseed) [inline src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala]: We should be able to call `coalesce` on RDDs with empty partitions.  This needs to get fixed.

File: src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala (hunk 23 above)
Comment (tpoterba) [inline src/test/scala/org/broadinstitute/hail/variant/vsm/VSMSuite.scala]: ah yup, used this to catch the empty RDD bug and forgot to change it back!

File: src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala (hunk 9 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala]: I'm good wth this.  You can call `c.clear()` instead of constructing another builder at `c = cbf()`.

**Comment by cseed** [comment]
OK, I think that's everything, back to 

I found this change hard to reason about.  Part of it was the layered changes.  Let's try to keep unrelated changes in unrelated commits.  Part of it is the inherent complexity of the Ordered stuff and Spark internals, although I don't have specific suggestions beyond my comments above.

**Comment by tpoterba** [comment]
Addressed comments, back to you.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedLeftJoinRDD.scala [hunk 25]
```diff
@@ -9,22 +9,23 @@ import org.broadinstitute.hail.utils.richUtils.RichPairIterator
...
+    context: TaskContext, k0: K): OrderedRDDIterator[PK, K, V] = {
+    val it = new OrderedRDDIterator(rdd, cachedPartitions, context)
     it.setPartition(rdd.orderedPartitioner.getPartition(k0))
     it
   }
 }
 
 class OrderedRDDIterator[PK, K, V](
   rdd: OrderedRDD[PK, K, V],
+  cachedPartitions: Array[Partition],
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedLeftJoinRDD.scala]: Did you not want to just rename this `partitions`?  Perhaps with a quick comment on why you can't just access `rdd.partitions`?

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala [hunk 26]
```diff
@@ -13,22 +13,24 @@ import scala.reflect.{ClassTag, classTag}
...
+  (implicit val kOk: OrderedKey[PK, K])
   extends Partitioner {
 
   import kOk.pkct
   import kOk.pkOrd
   import Ordering.Implicits._
 
-  require(rangeBounds.isEmpty ||
-    rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => left < right })
+  require(nPartitions == 0 || nPartitions == rangeBounds.length + 1,
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala]: Shouldn't the first condition be `(nPartitions == 0 && rangeBounds.isEmpty)`?

File: src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala (hunk 9 above)
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/BlockedRDD.scala]: Did you not want to throw this in?

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala [hunk 27]
```diff
@@ -83,18 +85,18 @@ case class OrderedPartitioner[PK, K](rangeBounds: Array[PK], ascending: Boolean
...
 object OrderedPartitioner {
   def empty[PK, K](implicit kOk: OrderedKey[PK, K]): OrderedPartitioner[PK, K] =
-    new OrderedPartitioner[PK, K](Array.empty(kOk.pkct))
+    new OrderedPartitioner[PK, K](Array.empty(kOk.pkct), 0)
 
   def read[PK, K](in: ObjectInputStream)(implicit kOk: OrderedKey[PK, K]): OrderedPartitioner[PK, K] = {
     val ascending = in.readBoolean()
     val rangeBounds = in.readObject().asInstanceOf[Array[PK]]
-    OrderedPartitioner(rangeBounds, ascending)
+    OrderedPartitioner(rangeBounds, rangeBounds.length + 1, ascending)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedPartitioner.scala]: Isn't this the same as the original `def nPartitions = rangeBounds.length + 1` bug?  You should read in the RDD in `VSM.read` and then pass in the number of partitions.

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 28]
```diff
@@ -42,82 +47,87 @@ object OrderedRDD {
...
+          (ARRAY_SORT, OrderedRDD(reorderedPartitionsRDD.mapPartitions { it =>
             it.toArray.sortBy(_._1).iterator
-          }, partitioner)
+          }, partitioner))
       }
     } else {
       info("Ordering unsorted dataset with network shuffle")
       val ranges = calculateKeyRanges(keys.map(kOk.project))
-      apply(rdd, ranges)
+      (SHUFFLE, shuffle(rdd, OrderedPartitioner(ranges, ranges.length + 1)))
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Shouldn't this be `rdd.partitions.length`?

File: src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala [hunk 29]
```diff
@@ -391,11 +338,11 @@ class OrderedRDD[PK, K, V] private(rdd: RDD[(K, V)], val orderedPartitioner: Ord
     assert(newPartEnd.zip(newPartEnd.tail).forall { case (i, inext) => i < inext })
 
     if (newPartEnd.length < maxPartitions)
-      warn(s"coalesced to ${newPartEnd.length} partitions, less than requested $maxPartitions")
-
-    val newRangeBounds = newPartEnd.init.map(orderedPartitioner.rangeBounds)
+      warn(s"coalesced to ${ newPartEnd.length } partitions, less than requested $maxPartitions")
 
-    new OrderedRDD[PK, K, V](new BlockedRDD(persisted, newPartEnd), OrderedPartitioner(newRangeBounds))
+    val newRangeBounds = Array.tabulate(newPartEnd.length - 1) { i => orderedPartitioner.rangeBounds(newPartEnd(i)) }
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/sparkextras/OrderedRDD.scala]: Does this do the same thing?  It doesn't seem like an improvement.

**Comment by cseed** [comment]
OK, I'm done.

**Comment by tpoterba** [comment]
Back to you!
